README.org
* A reasonable Emacs config
  :PROPERTIES:
  :CUSTOM_ID: a-reasonable-emacs-config
  :END:

This is my emacs configuration tree, continually used and tweaked since
2000, and it may be a good starting point for other Emacs users,
especially those who are web developers. These days it's somewhat geared
towards OS X, but it is known to also work on Linux and Windows.

Emacs itself comes with support for many programming languages. This
config adds improved defaults and extended support for the following:

-  Ruby / Ruby on Rails
-  CSS / LESS / SASS / SCSS
-  HAML / Markdown / Textile / ERB
-  Clojure (with Cider and nRepl)
-  Javascript / Coffeescript
-  Python
-  PHP
-  Haskell
-  Elm
-  Erlang
-  Common Lisp (with Slime)

In particular, there's a nice config for /autocompletion/ with
[[https://company-mode.github.io/][company]], and [[http://www.flycheck.org][flycheck]] is used to immediately highlight
syntax errors in Ruby, Python, Javascript, Haskell and a number of other
languages.

** Supported Emacs versions
   :PROPERTIES:
   :CUSTOM_ID: supported-emacs-versions
   :END:

The config should run on Emacs 23.3 or greater and is designed to
degrade smoothly - see the Travis build - but note that Emacs 24 and
above is required for an increasing number of key packages, including
=magit=, =company= and =flycheck=, so to get full you should use the
latest Emacs version available to you.

Some Windows users might need to follow
[[http://xn--9dbdkw.se/diary/how_to_enable_GnuTLS_for_Emacs_24_on_Windows/index.en.html][these
instructions]] to get TLS (ie. SSL) support included in their Emacs.

** Other requirements
   :PROPERTIES:
   :CUSTOM_ID: other-requirements
   :END:

To make the most of the programming language-specific support in this
config, further programs will likely be required, particularly those
that [[https://github.com/flycheck/flycheck][flycheck]] uses to provide
on-the-fly syntax checking.

** Installation
   :PROPERTIES:
   :CUSTOM_ID: installation
   :END:

To install, clone this repo to =~/.emacs.d=, i.e. ensure that the
=init.el= contained in this repo ends up at =~/.emacs.d/init.el=:

#+BEGIN_EXAMPLE
    git clone https://github.com/watkinsr/emacs.d.git ~/.emacs.d
#+END_EXAMPLE

Upon starting up Emacs for the first time, further third-party packages
will be automatically downloaded and installed. If you encounter any
errors at that stage, try restarting Emacs, and possibly running
=M-x package-refresh-contents= before doing so.

** Important note about =ido=
   :PROPERTIES:
   :CUSTOM_ID: important-note-about-ido
   :END:

This config enables =ido-mode= completion in the minibuffer wherever
possible, which might confuse you when trying to open files using
@@html:<kbd>@@C-x C-f@@html:</kbd>@@, e.g. when you want to open a
directory to use =dired= -- if you get stuck, use
@@html:<kbd>@@C-f@@html:</kbd>@@ to drop into the regular =find-file=
prompt. (You might want to customize the =ido-show-dot-for-dired=
variable if this is an issue for you.)

** Updates
   :PROPERTIES:
   :CUSTOM_ID: updates
   :END:

Update the config with =git pull=. You'll probably also want/need to
update the third-party packages regularly too:

@@html:<kbd>@@M-x package-list-packages@@html:</kbd>@@, then
@@html:<kbd>@@U@@html:</kbd>@@ followed by
@@html:<kbd>@@x@@html:</kbd>@@.

You should usually restart Emacs after pulling changes or updating
packages so that they can take effect. Emacs should usually restore your
working buffers when you restart due to this configuration's use of the
=desktop= and =session= packages.

** Adding your own customization
   :PROPERTIES:
   :CUSTOM_ID: adding-your-own-customization
   :END:

To add your own customization, use @@html:<kbd>@@M-x
customize@@html:</kbd>@@ and/or create a file
=~/.emacs.d/lisp/init-local.el= which looks like this:

#+BEGIN_EXAMPLE
    ... your code here ...

    (provide 'init-local)
#+END_EXAMPLE

If you need initialisation code which executes earlier in the startup
process, you can also create an =~/.emacs.d/lisp/init-preload-local.el=
file.

If you plan to customize things more extensively, you should probably
just fork the repo and hack away at the config to make it your own!
Remember to regularly merge in changes from this repo, so that your
config remains compatible with the latest package and Emacs versions.

/Please note that I cannot provide support for customised versions of
this configuration./

** Similar configs
   :PROPERTIES:
   :CUSTOM_ID: similar-configs
   :END:

You might also want to check out =emacs-starter-kit= and =prelude=.


eshell/history
clear

sudo rm -rf /usr/lib/python3.6/
sudo rm -rf /usr/lib/python2.7
ls
clear

clear

ls
clear

scala
sudo pacman -S scala
sudo pacman -S yaourt
sudo pacman -Syu
sudo pacman -S yaourt
sudo pacman -Syu
sudo pacman -S yaourt
sudo pacman -S yaourt
sudo pacman -S package-query
cd Downloads/
ls
tar -xvzf yaourt.tar.gz 
cd yaourt/
makepkg -si
ls
cd ..
ls
untar -xvzf scala-2.12.1.tgz 
tar -xvzf scala-2.12.1.tgz 
cd scala-2.12.1/
ls
makepkg -A
cd ..
ls
rm -rf scala-2.12.1/
ls
clear

ls
sudo mkdir /usr/lib/scala
tar -xf scala-2.12.1.tgz -C /usr/lib/scala
sudo tar -xf scala-2.12.1.tgz -C /usr/lib/scala
rm scala-2.12.1.tgz
sudo ln -s /usr/lib/scala/scala-2.12.1/bin/scala /bin/scala
sudo ln -s /usr/lib/scala/scala-2.12.1/bin/scalac /bin/scalac
scala --version
sudo pacman -S java
sudo pacman -S jdk
sudo pacman -S jre8-openjdk
scala --version
scala
scala
scala
java -version
scala
sudo pacman -R openjre-jdk8
sudo pacman -R jdk8-openjre
sudo pacman -R jre8-openjdk
ls
ls
sudo mkdir /usr/lib/jvm
sudo tar -xf jdk-8u121-linux-arm64-vfp-hflt.tar.gz -C /usr/lib/jvm
mv /usr/lib/jvm/jdk1.8.0_121 /usr/lib/jvm/java-8-oracle
sudo mv /usr/lib/jvm/jdk1.8.0_121 /usr/lib/jvm/java-8-oracle
rm jdk-8u121-linux-arm64-vfp-hflt.tar.gz 
ln -s /usr/lib/jvm/java-8-oracle/bin/java /bin/java
ln -s /usr/lib/jvm/java-8-oracle/bin/java /bin/java
sudo rm /bin/java
ln -s /usr/lib/jvm/java-8-oracle/bin/java /bin/java
sudo ln -s /usr/lib/jvm/java-8-oracle/bin/java /bin/java
sudo rm /bin/javac
sudo ln -s /usr/lib/jvm/java-8-oracle/bin/javac /bin/javac
java -version
ls /usr/lib/jvm
sudo rm -rf /usr/lib/java-8*
ls /usr/lib/jvm
sudo rm -rf /usr/lib/jvm/java-8*
ls /usr/lib/jvm
sudo rm -rf /usr/lib/jvm/*
ls
sudo tar -xf jdk-8u121-linux-arm32-vfp-hflt.tar.gz /usr/lib/jvm
ls /usr/lib/jvm
ls /usr/lib
ls /usr/lib/jvm
cd /usr/lib/jvm
cd
sudo tar -xf jdk-8u121-linux-arm32-vfp-hflt.tar.gz /usr/lib/jvm
ls
cd Downloads/
sudo tar -xf jdk-8u121-linux-arm32-vfp-hflt.tar.gz /usr/lib/jvm
sudo tar -xf jdk-8u121-linux-arm32-vfp-hflt.tar.gz -C /usr/lib/jvm
sudo mv /usr/lib/jvm/jdk1.8.0_121 /usr/lib/jvm/java-8-oracle
rm jdk-8u121-linux-arm32-vfp-hflt.tar.gz 
sudo ln -s /usr/lib/jvm/java-8-oracle/bin/javac /bin/javac
sudo rm /bin/javac
sudo rm /bin/java
sudo ln -s /usr/lib/jvm/java-8-oracle/bin/javac /bin/javac
sudo ln -s /usr/lib/jvm/java-8-oracle/bin/java /bin/java
java -version
scala
localectl list-keymaps | grep -i colemak
localectl set-keymap --no-convert colemak
locatectl status
localectl status
reboot
cd
pactl
pactl load-module --help
pactl load-module 
pactl load-module 
pactl load-module module-simple-protocol-tcp
pactl load-module module-simple-protocol-tcp rate=48000 format=s161e channels=2 source=alsa record=true port=8000
pactl load-module module-simple-protocol-tcp rate=48000 format=s161e channels=2 source= record=true port=8000
pactl load-module module-simple-protocol-tcp rate=48000 format=s161e channels=2 record=true port=8000
pactl load-module module-simple-protocol-tcp rate=48000 format=s161e channels=2 record=true port=8000
pactl list | grep Name
pactl load-module module-simple-protocol-tcp rate=48000 format=s161e channels=2 source=alsa_output.0.analog-stereo record=true port=8000
pactl load-module module rate=48000 format=s161e channels=2 source=alsa_output.0.analog-stereo record=true port=8000
pactl load-module module-simple-protocol-tcp
pactl load-module module-simple-protocol-tcp rate=48000 format=s161e
pactl load-module module-simple-protocol-tcp rate=48000
mpv https://www.youtube.com/watch?v=6xaaeop7gJ8 &
mpv http://www.sextvx.com/de/video/442597/kiki-vidis-threesome
clear

screen


init.el
;;; This file bootstraps the configuration, which is divided into
;;; a number of other files.

(let ((minver "23.3"))
  (when (version<= emacs-version minver)
    (error "Your Emacs is too old -- this config requires v%s or higher" minver)))
(when (version<= emacs-version "24")
  (message "Your Emacs is old, and some functionality in this config will be disabled. Please upgrade if possible."))

(add-to-list 'load-path (expand-file-name "lisp" user-emacs-directory))
(require 'init-benchmarking) ;; Measure startup time

(defconst *spell-check-support-enabled* nil) ;; Enable with t if you prefer
(defconst *is-a-mac* (eq system-type 'darwin))

;;----------------------------------------------------------------------------
;; Temporarily reduce garbage collection during startup
;;----------------------------------------------------------------------------
(defconst sanityinc/initial-gc-cons-threshold gc-cons-threshold
  "Initial value of `gc-cons-threshold' at start-up time.")
(setq gc-cons-threshold (* 128 1024 1024))
(add-hook 'after-init-hook
          (lambda () (setq gc-cons-threshold sanityinc/initial-gc-cons-threshold)))

;;----------------------------------------------------------------------------
;; Bootstrap config
;;----------------------------------------------------------------------------
(setq custom-file (expand-file-name "custom.el" user-emacs-directory))
(require 'init-compat)
(require 'init-utils)
(require 'init-site-lisp) ;; Must come before elpa, as it may provide package.el
;; Calls (package-initialize)
(require 'init-elpa)      ;; Machinery for installing required packages
(require 'init-exec-path) ;; Set up $PATH

;;----------------------------------------------------------------------------
;; Allow users to provide an optional "init-preload-local.el"
;;----------------------------------------------------------------------------
(require 'init-preload-local nil t)

;;----------------------------------------------------------------------------
;; Load configs for specific features and modes
;;----------------------------------------------------------------------------

(require-package 'wgrep)
(require-package 'project-local-variables)
(require-package 'diminish)
(require-package 'scratch)
(require-package 'mwe-log-commands)

(require 'init-frame-hooks)
(require 'init-xterm)
(require 'init-color-theme)
(require 'init-emacs-w3m)
;; (require 'init-themes)
(require 'init-osx-keys)
(require 'init-gui-frames)
(require 'init-dired)
(require 'init-isearch)
(require 'init-grep)
(require 'init-uniquify)
(require 'init-ibuffer)
(require 'init-flycheck)

(require 'init-recentf)
(require 'init-smex)
;; If you really prefer ido to ivy, change the comments below. I will
;; likely remove the ido config in due course, though.
;; (require 'init-ido)
(require 'init-ivy)
(require 'init-hippie-expand)
(require 'init-company)
(require 'init-windows)
(require 'init-sessions)
(require 'init-fonts)
(require 'init-mmm)

(require 'init-editing-utils)
(require 'init-whitespace)
(require 'init-fci)

(require 'init-vc)
(require 'init-darcs)
(require 'init-git)
(require 'init-github)

(require 'init-projectile)

(require 'init-compile)
(require 'init-crontab)
(require 'init-textile)
(require 'init-markdown)
(require 'init-csv)
(require 'init-erlang)
(require 'init-javascript)
(require 'init-php)
(require 'init-org)
(require 'init-nxml)
(require 'init-html)
(require 'init-css)
(require 'init-haml)
(require 'init-python-mode)
(unless (version<= emacs-version "24.3")
  (require 'init-haskell))
(require 'init-elm)
(require 'init-ruby-mode)
(require 'init-rails)
(require 'init-sql)

(require 'init-paredit)
(require 'init-lisp)
(require 'init-slime)
(unless (version<= emacs-version "24.2")
  (require 'init-clojure)
  (require 'init-clojure-cider))
(require 'init-common-lisp)

(when *spell-check-support-enabled*
  (require 'init-spelling))

(require 'init-misc)

(require 'init-folding)
(require 'init-dash)
(require 'init-ledger)

;;---------------------------------------------------------------------------
;; RYAN'S EXTRA STUFF
;;---------------------------------------------------------------------------
(require 'init-evil)
(require 'init-gnus)
(require 'init-workgroups2)
(require 'init-elscreen)
(workgroups-mode 1)

(require 'init-linum-mode)
(require 'init-erc)

;; TODO :: Seperate to init-go
(add-to-list 'load-path "~/go-autocomplete/")
(require 'go-autocomplete)
(require 'auto-complete-config)
(ac-config-default)

(require 'go-eldoc)
(add-hook 'go-mode-hook 'go-eldoc-setup)

;; SCALA
(require 'ensime)

;; REST
(require 'restclient)

;; UTF-8 TERM encoding
(setenv "LANG" "en_US.UTF-8")

;; PDF-TOOLS
(pdf-tools-install)

(blink-cursor-mode 0)

;; INDENT
(setq tab-width 4)

;; FRAME MANAGEMENT
(defun ido-switch-frame ()
  "Switch frame with ido"
  (interactive)
  (when (not (minibufferp))
    (let* (
           (frames (frame-list))
           (frame-to (ido-completing-read "Select Frame:  "
                                          (mapcar (lambda (frame) (frame-parameter frame 'name)) frames))))
      (catch 'break
        (while frames
          (let ((frame (car frames)))
            (if (equal (frame-parameter frame 'name) frame-to)
                (throw 'break (select-frame-set-input-focus frame))
              (setq frames (cdr frames)))))))))

(global-set-key (kbd "C-x 4 z") 'ido-switch-frame)

;; Useful for quick feedback from functions
(global-set-key (kbd "C-x 4 e") 'eval-region)

;; Color-theme hack, load-theme somehow lost between Purcell and my fork
(load-theme 'molokai t)

(defun paste-from-x-clipboard ()
  (interactive)
  (shell-command "pbpaste" 1))

(global-set-key (kbd "M-y") 'paste-from-x-clipboard)

;; AucTeX setup
(setq TeX-auto-save t)
(setq TeX-parse-self t)
(setq-default TeX-master nil)

;; Open urls with icecat
(setq browse-url-browser-function 'browse-url-generic
      browse-url-generic-program "icecat")

;; Tramp setup...
(add-to-list 'load-path "~/.emacs.d/tramp-2.2.7/lisp")
(require 'tramp)

;; Make sure we work on remote guixsd machines :)
;; probably only helps if you start on a guixsd machine..!
(setq tramp-remote-path
      (append tramp-remote-path
              '(tramp-own-remote-path)))

(global-linum-mode 0)

;; MISC stuff from Purcell
(require-package 'gnuplot)
(require-package 'lua-mode)
(require-package 'htmlize)
(require-package 'dsvn)
(when *is-a-mac*
  (require-package 'osx-location))
(require-package 'regex-tool)

;;----------------------------------------------------------------------------
;; Allow access from emacsclient
;;----------------------------------------------------------------------------
(require 'server)
(unless (server-running-p)
  (server-start))


;;----------------------------------------------------------------------------
;; Variables configured via the interactive 'customize' interface
;;----------------------------------------------------------------------------
(when (file-exists-p custom-file)
  (load custom-file))


;;----------------------------------------------------------------------------
;; Allow users to provide an optional "init-local" containing personal settings
;;----------------------------------------------------------------------------
(require 'init-local nil t)


;;----------------------------------------------------------------------------
;; Locales (setting them earlier in this file doesn't work in X)
;;----------------------------------------------------------------------------
(require 'init-locales)

(provide 'init)

;; Local Variables:
;; coding: utf-8
;; no-byte-compile: t
;; End:
